lib/miio/RetryWrapper.js

```diff
@@ -93,23 +94,23 @@ class RetryWrapper {
      * @returns {Promise<object>}
      */
     async sendMessage(method, args = [], options = {}) {
-        if (!this.miioSocket.stamp.isValid()) {
-            await this.handshake(false);
-
-            return await this.sendMessage(method, args, options);
-        }
-
         const msg = {method: method, params: args};
 
         try {
+            if (!this.miioSocket.stamp.isValid()) {
+                // Trigger retry after performing new handshake
+                // noinspection ExceptionCaughtLocallyJS
+                throw new MiioInvalidStampError();
+            }
+
```

Comment:
Sometimes, if I try to **re**start Valutudo when the robot was already connected, the stamp isn't valid and some code later needs to be reached, but I'm still haven't found it.

This is a workaround, it still prints a bunch of errors later on `newToken.equals()` saying that there's no such method, but eventually it works so I'd keep it as a quick and dirty workaround temporarily until the viomi implementation is in a good shape.